diffkemp/simpll/Logger.h

```diff
@@ -0,0 +1,113 @@
+//===---------------- Logger.h - Logging debug information ----------------===//
+//
+//       SimpLL - Program simplifier for analysis of semantic difference      //
+//
+// This file is published under Apache 2.0 license. See LICENSE for details.
+// Author: Tomas Kucma, xkucma00@vutbr.cz
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the declaration of the Logger class that provides
+/// debug logging functions.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef DIFFKEMP_SIMPLL_LOGGER_H
+#define DIFFKEMP_SIMPLL_LOGGER_H
+
+#include <llvm/Support/Debug.h>
+#include <llvm/Support/raw_ostream.h>
+#include <vector>
+
+#define LOGGER_BASE_LEVEL DEBUG_SIMPLL_VERBOSE
+#define LOGGER_FORCE_LEVEL DEBUG_SIMPLL_VERBOSE_EXTRA
+
+// Prepare a log messsage for (potential) future logging and create context for
+// logging its children. Must be later followed by LOG_KEEP or RETURN_WITH_LOG
+// (or their _FORCE variants)
+#define PREP_LOG(msg)                                                          \
+    DEBUG_WITH_TYPE(LOGGER_BASE_LEVEL, logger.prep_log() msg;                  \
+                    logger.prep_context();)
+
+// Based on given value (interpreted as bool), either keep or erase prepared
+// message and its children. Exectuing keep on a lowest level message leads to
+// automatically printing it and all its children, clearing the message buffer.
+#define LOG_KEEP(keep) DEBUG_WITH_TYPE(LOGGER_BASE_LEVEL, logger.log(x))
```

Comment:
Is the `keep` argument here actually used somewhere? It looks to me that it is not, and `x`, which is passed as the `keep` value, is just passed from the context of the parent macro, in that case the argument would be redundant.

Reply:
Thanks for pointing that out, I meant to change the x to keep for this macro. It's meant to have it as an argument, so it can potentially be used instead of RETURN_WITH_LOG in cases where returning the given value isn't desired.